truffle/src/main/java/org/jruby/truffle/nodes/ext/psych/PsychEmitterNodes.java

```diff
@@ -0,0 +1,493 @@
+            options.setIndent((int) ruby("options_set.indentation", "options_set", optionsSet));
+            Layouts.PSYCH_EMITTER.setOptions(emitter, options);
+
+            Layouts.PSYCH_EMITTER.setIo(emitter, io);
+
+            return nil();
+        }
+
+    }
+
+    @CoreMethod(names = "start_stream", required = 1)
+    public abstract static class StartStreamNode extends CoreMethodArrayArgumentsNode {
+
+        public StartStreamNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        @CompilerDirectives.TruffleBoundary
+        @Specialization
+        public DynamicObject startStream(DynamicObject emitter, int encoding) {
+            initEmitter(getContext(), emitter, encoding);
+
+            StreamStartEvent event = new StreamStartEvent(NULL_MARK, NULL_MARK);
+
+            emit(getContext(), emitter, event);
+
+            return emitter;
+        }
+
+    }
+
+
+    @CoreMethod(names = "end_stream")
+    public abstract static class EndStreamNode extends CoreMethodArrayArgumentsNode {
+
+        public EndStreamNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        @CompilerDirectives.TruffleBoundary
+        @Specialization
+        public DynamicObject endStream(DynamicObject emitter) {
+            emit(getContext(), emitter, new StreamEndEvent(NULL_MARK, NULL_MARK));
+            return emitter;
+        }
+
+    }
+
+
+    @CoreMethod(names = "start_document", required = 3)
+    public abstract static class StartDocumentNode extends CoreMethodArrayArgumentsNode {
+
+        public StartDocumentNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        @CompilerDirectives.TruffleBoundary
+        @Specialization(guards = {"isRubyArray(_version)", "isRubyArray(tags)"})
+        public DynamicObject startDocument(DynamicObject emitter, DynamicObject _version, DynamicObject tags, boolean implicit) {
+            DumperOptions.Version version = null;
+            boolean implicitBool = implicit;
+            Map<String, String> tagsMap = null;
+
+            Object[] versionAry = ArrayOperations.toObjectArray(_version);
+            if (versionAry.length == 2) {
+                int versionInt0 = (int)versionAry[0];
+                int versionInt1 = (int)versionAry[1];
+
+                if (versionInt0 == 1) {
+                    if (versionInt1 == 0) {
+                        version = DumperOptions.Version.V1_0;
+                    } else if (versionInt1 == 1) {
+                        version = DumperOptions.Version.V1_1;
+                    }
+                }
+                if (version == null) {
+                    throw new UnsupportedOperationException();
+                    //throw context.runtime.newArgumentError("invalid YAML version: " + versionAry);
```

Comment:
Sorry, it's not leftover junk but things I haven't bothered to implement yet. I copied-and-pasted the JRuby code to get going, and this was things I hadn't gotten around to yet. They should be to-dos really. I'll fix that.